modules/core/src/main/kotlin/datamaintain/core/step/Scanner.kt

```diff
@@ -3,12 +3,39 @@ package datamaintain.core.step
 import datamaintain.core.Context
 import datamaintain.core.script.FileScript
 import datamaintain.core.script.ScriptWithContent
+import datamaintain.core.script.Tag
+import java.io.File
+import java.nio.file.Path
 
 class Scanner(private val context: Context) {
-    fun scan(): List<ScriptWithContent> = context.config.path.toFile().walk()
-            .filter { it.isFile }
-            .map {
-                FileScript(it.toPath(), context.config.identifierRegex)
-            }
-            .sortedBy { it.name }.toList()
-}
\ No newline at end of file
+    fun scan(): List<ScriptWithContent> {
+        val rootFolder: File = context.config.path.toFile()
+
+        return rootFolder.walk()
+                .filter { it.isFile }
+                .map {
+                    val tags = mutableSetOf<Tag>()
+
+                    if (context.config.doesCreateTagsFromFolder) {
+                        tags.addAll(buildTagsFromFolder(rootFolder, it));
+                    }
+
+                    FileScript(it.toPath(), context.config.identifierRegex, tags.toSet())
+                }
+                .sortedBy { it.name }.toList()
+    }
+
+    private fun buildTagsFromFolder(rootFolder: File, it: File): Set<Tag> {
+        val parentTags = mutableSetOf<Tag>()
+        val relativePath = it.relativeTo(rootFolder).toPath()
+
+        var parent: Path? = relativePath.parent
+
+        while (parent != null) {
+            parentTags.add(Tag(parent.toFile().name))
+            parent = parent.parent
+        }
+
+        return parentTags
+    }
+}
```

Comment:
Since the tree view is done, I don't see the point in looking for files parents when they are already known. Why not do a depth-first search?

Reply:
I think I understand what you want but I would like to discuss that with you.